File: learning/leetcode/LC206.py
# -*- coding: utf-8 -*-
# LeetCode 206 - 反转链表
# 运行: python learning/leetcode/LC206.py
#
# 题目: 把 1 -> 2 -> 3 -> 4 -> None  变成  4 -> 3 -> 2 -> 1 -> None
#
# ============================================================
# 第一步: 什么是链表？（从零讲）
# ============================================================
#
# 链表就像「火车」——每节车厢（Node）装着货（data），
# 但它只知道下一节车厢是谁（next），不知道整辆车的样子。
#
#   车厢1       车厢2       车厢3       终点
#   [1|●]  →   [2|●]  →   [3|●]  →   None
#    ↑ 车厢号    ↑ 车厢号    ↑ 车厢号    ↑ 没了=结尾
#
#   ● = next指针 = 指向下一节车厢


class Node:
    """链表的一节「车厢」"""
    def __init__(self, val):
        self.val = val    # 车厢里装的货（数据）
        self.next = None  # 指向下一节车厢（默认没有下一节）

    def __repr__(self):
        return f"Node({self.val})"


# ============================================================
# 第二步: 徒手造一条链表
# ============================================================

# 造 3 节车厢: 1 → 2 → 3 → None
a = Node(1)
b = Node(2)
c = Node(3)

# 把它们连起来
a.next = b   # 1 后面是 2
b.next = c   # 2 后面是 3
# c.next 默认为 None，表示终点

# 这个 a 就是「头节点」，顺着 a 一路 .next 就能走完整个链表
print("原链表:")
node = a
while node:
    print(f"  {node.val}", end="")
    if node.next:
        print(" → ", end="")
    node = node.next
print(" → None")
# 输出: 1 → 2 → 3 → None


# ============================================================
# 第三步: 反转链表（核心）
# ============================================================
#
# 目标: 把 a(1)→b(2)→c(3)→None 变成 c(3)→b(2)→a(1)→None
#
# 关键想法（一句话）:
#   让每个节点的 next 掉头，指向前一个节点！
#
# 准备 3 根「手指」指着不同的车厢:
#   prev  ← 已经掉好头的部分的头
#   curr  ← 正在掉头的那个车厢
#   nxt   ← 还没处理的下一节车厢（提前记好，别弄丢了）
#
# 图解过程（读代码时对照着看）:
#
#   初始状态:
#         prev=None    curr=a(1)     nxt=??
#         (空的)       ↓             ↓
#                      [1|●]→[2|●]→[3|●]→None
#
#   第1步: 记下下一节 nxt = curr.next  (拍照存证，免得回头找不到了)
#   第2步: 让当前车厢掉头 curr.next = prev  (1.next 从指向2 变成指向None)
#   第3步: prev 往前走一步  prev = curr
#   第4步: curr 往前走一步  curr = nxt
#
#         完成后:
#         None←[1]    prev=1    curr=2    nxt=2
#
#   然后循环重复上面的4步...直到 curr 走到 None（链表尽头）
#
#   最终: None←[1]←[2]←[3]  ← 全部掉头完成
#                       ↑ prev 指向新的头


def reverse_list(head):
    """反转链表，返回新的头节点"""
    prev = None   # 上一个节点（一开始没有，就是 None）
    curr = head   # 当前节点（从头开始）

    while curr is not None:  # 只要还没走完
        nxt = curr.next      # ① 拍照：记下下一个节点
        curr.next = prev     # ② 掉头：让当前节点指向前一个
        prev = curr          # ③ 前进一步
        curr = nxt           # ④ 前进一步

    # 循环结束时 curr = None，prev 就是新的头
    return prev


# 测试
new_head = reverse_list(a)

print("\n反转后:")
node = new_head
while node:
    print(f"  {node.val}", end="")
    if node.next:
        print(" → ", end="")
    node = node.next
print(" → None")
# 期望输出: 3 → 2 → 1 → None


# ============================================================
# 第四步: LeetCode 官方格式
# ============================================================
#
# LeetCode 给你的是头节点 head，你要返回新的头节点。
# 上面那个 reverse_list 就是 LeetCode 能直接提交的答案。
# 下面是 LeetCode 上的最终版本：

class Solution:
    def reverseList(self, head):
        prev = None
        curr = head
        while curr:
            nxt = curr.next   # 存好下一个
            curr.next = prev  # 掉头
            prev = curr       # 往前走
            curr = nxt        # 往前走
        return prev


# ============================================================
# 第五步: 慢慢看每一步发生了什么
# ============================================================
print("\n" + "=" * 50)
print("慢动作回放：每一步的细节")
print("=" * 50)

# 重新造一条链表
a2 = Node(1)
b2 = Node(2)
c2 = Node(3)
a2.next = b2
b2.next = c2

prev = None
curr = a2
step = 0

while curr:
    step += 1
    nxt = curr.next  # 拍照存证
    print(f"\n第{step}步:")
    print(f"  当前车厢 curr = {curr.val}")
    print(f"  它的下一节 nxt = {nxt.val if nxt else 'None'} ← 提前记下，别丢了")
    print(f"  掉头前: {curr.val}.next → {curr.next.val if curr.next else 'None'}")
    curr.next = prev  # 掉头
    print(f"  掉头后: {curr.val}.next → {prev.val if prev else 'None'} ← 指回去了！")
    prev = curr
    curr = nxt
    print(f"  移动后: prev={prev.val}, curr={curr.val if curr else 'None'}")

print(f"\n反转完成！新头是 prev = {prev.val}")
print("顺着新头走一遍验证一下...")

node = prev
parts = []
while node:
    parts.append(str(node.val))
    node = node.next
print("  " + " -> ".join(parts) + " -> None [OK]")


# ============================================================
# 记忆口诀
# ============================================================
print("\n" + "=" * 50)
print("四句口诀，记住反转链表:")
print("  记下下一个（nxt = curr.next）")
print("  掉头指向前（curr.next = prev）")
print("  前进一步（prev = curr）")
print("  再进一步（curr = nxt）")
print("=" * 50)

File: learning/leetcode/LC206.cpp
/*
 * LeetCode 206 - Reverse Linked List (反转链表)
 *
 * 题目: Input:  1 -> 2 -> 3 -> 4 -> 5 -> nullptr
 *       Output: 5 -> 4 -> 3 -> 2 -> 1 -> nullptr
 *
 * 编译: /c/msys64/mingw64/bin/g++ -std=c++17 -o lc206.exe LC206.cpp
 * 运行: ./lc206.exe
 *
 * ============================================================
 * 第一讲: 什么是链表？
 * ============================================================
 *
 * 链表就是「火车」。
 *
 *   [1|->]     [2|->]     [3|->]     [4|->]     [5|  ]
 *   车厢         车厢        车厢       车厢        车厢
 *   ^data       ^data       ^data      ^data       ^data
 *       ^next       ^next       ^next       ^next      ^next = nullptr (终点)
 *
 * 每节车厢（Node）只知道自己:
 *   1. 装着什么东西 (val)
 *   2. 下一节车厢在哪 (next 指针)
 *
 * 关键区别:
 *   Python: self.val, self.next, None
 *   C++:    val, next (是 node* 类型的指针), nullptr
 *           要用 new 创建节点，用指针来指
 *
 * ============================================================
 * 第二讲: C++ 里的指针是什么？
 * ============================================================
 *
 * 指针 = 一张写着「某个东西的位置」的便签纸
 *
 *   int x = 42;       // x 是一块内存，装着 42
 *   int* p = &x;      // p 是一张便签纸，写着 "x 在 XXXX 号位置"
 *   cout << *p;       // *p = "我去便签上写的地址把东西取出来" → 42
 *   cout << p;        // p  = "便签上写的地址本身" → 0x7ffe...
 *
 * 在链表里:
 *   Node* node = new Node(1);  // 火车站调度员说"新车厢1号停到D-3站台"
 *   node->val = 1;             // 去D-3站台，打开车厢，放1进去
 *   node->next = nextNode;     // 把下一节车厢的站台号写在车门上
 *
 *   -> 就是「先去便签上的地址，再进门拿东西」，Python 里的 . 的指针版
 *
 * ============================================================
 */

#include <iostream>
#include <vector>
using namespace std;


// --------- 定义车厢 ----------
struct ListNode {
    int val;            // 车厢里的货
    ListNode *next;     // 下一节车厢的站台号（指针）

    // C++ 的「造车厢说明书」(构造函数)
    // Python: def __init__(self, val=0, next=None)
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *n) : val(x), next(n) {}
};


// ============================================================
// 第三讲: 造一条链表（和 Python 对比）
// ============================================================

// Python 版:
//   a = Node(1); b = Node(2); c = Node(3)
//   a.next = b; b.next = c

ListNode* makeList(vector<int> nums) {
    // 虚拟头节点(dummy)技巧: 先造一节空车厢，最后去掉
    // 好处: 不用单独处理「链表还是空的」情况
    ListNode dummy(0);          // 空车厢，不开出门，只是造车指挥所
    ListNode* tail = &dummy;   // tail 指着最后一节车厢（一开始只有指挥所）

    for (int val : nums) {
        tail->next = new ListNode(val);   // 新车厢挂上去
        tail = tail->next;                // tail 移到新车厢
    }

    return dummy.next;  // 返回真正的第一节车厢（甩掉指挥所）
}

// 打印链表
void printList(ListNode* head) {
    if (!head) {
        cout << "[empty]" << endl;
        return;
    }
    ListNode* curr = head;
    while (curr) {
        cout << curr->val;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << " -> nullptr" << endl;
}


// ============================================================
// 第四讲: 反转链表（核心算法）
// ============================================================
/*
 * 一句话: 让每个节点的箭头（next）反着指
 *
 * 看下面这个图（从左向右读）:
 *
 *   nullptr <- 1 <- 2 <- 3   本质就是 node.next 方向变了
 *
 * 怎么做? 需要 3 根手指同时指着不同地方:
 *
 *   prev  ← 已经掉好头的那些车厢的头
 *   curr  ← 正在掉头的车厢
 *   nxt   ← 下一节还没处理的车厢（提前记住，免得弄丢了）
 *
 * 初始状态:
 *   prev = nullptr     (还没开始掉头，所以「已经掉好头的链」是空的)
 *   curr = head        (从第一节开始)
 *
 * 循环每次做 4 件事:
 *   1. nxt = curr->next      拍照: 把下一节提前记下
 *   2. curr->next = prev     掉头: 让当前车厢的箭头指向后面
 *   3. prev = curr           前移
 *   4. curr = nxt            前移
 *
 * 图解一步:
 *
 *   掉头前:   prev=nullptr    curr=[1|->2]    nxt=??
 *            nullptr          [1|->] [2|->] [3|->] nullptr
 *
 *   第1步:    nxt = [2]       ← 拍照，记住下一节位置
 *   第2步:    1 -> nullptr    ← 反向！
 *            nullptr <- [1]      [2|->] [3|->] nullptr
 *
 *   第3步:    prev = [1]
 *   第4步:    curr = [2]
 *
 *   重复上面 4 步，直到 curr == nullptr（走到尽头）
 *   最后 prev 就是新的头节点
 */

ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;   // 已经掉头完成的部分的头
    ListNode* curr = head;      // 正在处理的节点

    while (curr != nullptr) {
        ListNode* nxt = curr->next;   // 1. 拍照存好下一个
        curr->next = prev;             // 2. 掉头！指向前一个
        prev = curr;                   // 3. prev 前进
        curr = nxt;                    // 4. curr 前进
    }

    return prev;  // 新的头
}


// ============================================================
// 第五讲: LeetCode 提交格式
// ============================================================

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (curr != nullptr) {
            ListNode* nxt = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
        }

        return prev;
    }
};


// ============================================================
// 第六讲: 慢动作回放
// ============================================================
void slowMotionDemo() {
    cout << "\n===========================================" << endl;
    cout << "慢动作回放: 反转 [1, 2, 3]" << endl;
    cout << "===========================================" << endl;

    ListNode* a = new ListNode(1);
    ListNode* b = new ListNode(2);
    ListNode* c = new ListNode(3);
    a->next = b;
    b->next = c;

    cout << "初始链表: ";
    printList(a);

    ListNode* prev = nullptr;
    ListNode* curr = a;
    int step = 0;

    while (curr) {
        step++;
        ListNode* nxt = curr->next;  // 1. 拍照
        cout << "\n第" << step << "步:" << endl;
        cout << "  curr = " << curr->val << endl;
        cout << "  它的 next = " << (nxt ? to_string(nxt->val) : "nullptr");
        cout << "  ← 拍照存好，别丢了" << endl;
        cout << "  掉头前: " << curr->val << ".next -> ";
        cout << (curr->next ? to_string(curr->next->val) : "nullptr") << endl;

        curr->next = prev;           // 2. 掉头

        cout << "  掉头后: " << curr->val << ".next -> ";
        cout << (prev ? to_string(prev->val) : "nullptr");
        cout << "  ← 箭头反过来了！" << endl;

        prev = curr;                 // 3. 前进
        curr = nxt;                  // 4. 前进

        cout << "  移动后: prev=" << prev->val;
        cout << ", curr=" << (curr ? to_string(curr->val) : "nullptr") << endl;
    }

    cout << "\n反转完成! 新头 = " << prev->val << endl;
    cout << "验证: ";
    printList(prev);

    // 清理内存（C++ 需要自己删，Python 不需要）
    ListNode* node = prev;
    while (node) {
        ListNode* tmp = node;
        node = node->next;
        delete tmp;
    }
}


// ============================================================
// 口诀
// ============================================================

int main() {
    cout << "===========================================" << endl;
    cout << "LeetCode 206 -- Reverse Linked List" << endl;
    cout << "===========================================" << endl;

    // 造一条链表: [1, 2, 3, 4, 5]
    ListNode* head = makeList({1, 2, 3, 4, 5});

    cout << "\n原链表: ";
    printList(head);

    // 反转
    head = reverseList(head);

    cout << "反转后: ";
    printList(head);

    // 慢动作回放
    slowMotionDemo();

    // 清理内存
    while (head) {
        ListNode* tmp = head;
        head = head->next;
        delete tmp;
    }

    cout << "\n===========================================" << endl;
    cout << "四句口诀:" << endl;
    cout << "  拍照存好下一个 (nxt = curr->next)" << endl;
    cout << "  箭头反着指     (curr->next = prev)" << endl;
    cout << "  挪一步         (prev = curr)" << endl;
    cout << "  再挪一步       (curr = nxt)" << endl;
    cout << "===========================================" << endl;

    return 0;
}
